nano/node/scheduler/optimistic.cpp

```diff
@@ -137,7 +135,7 @@ void nano::scheduler::optimistic::run ()
 
 				lock.unlock ();
 
-				run_one (transaction, candidate);
+				run_one (ledger.store.tx_begin_read (), candidate);
```

Comment:
Was this tested to actually improve things? I know that for a write transactions it's better to have a longer running transaction and the difference is huge.

Reply:
The main purpose is to eliminate holding transactions while making callbacks or notifying observers. I added a PR comment linking to the PR that adds a shared_lock to ledger transactions to guard memory objects.

Reply:
Okay, the intention is good but I'm not sure it's the way to fix it, especially since this would imply many other components need to be changed too.

Reply:
A call to `transaction.refresh_is_needed` might be necessary, but requiring starting a fresh transaction for each iteration should never be necessary in a properly designed system.

Reply:
There are options to batch these options if there's a performance problem, which we don't know if there is.

This loop could: copy the candidates, unlock, create db transaction, and run operation on each copied item, lock.

Reply:
So the problem is with keeping long lived transaction while holding its own mutex, is this correct? I might have misunderstood the problem.